### ISSUE TITLE
Certain image layout formats cannot be mapped

### ISSUE THREAD

**Issue by BearishSun** [LukasBanana/XShaderCompiler#48]
I have just ran into an issue that I'm not sure how to solve using pure HLSL as input. 

GLSL requires `image*` types to be prefixed with a layout format qualifier, if being used in loads or atomics (`layout(rgba32f)`). The problem is that these formats need to correspond to the actual data in the buffer - meaning if your buffer contains 8-bit normalized float values, the qualifier should be `layout(rgba8)`.

But in HLSL the buffer type (e.g. `RWBuffer<float4>`) represents the final type visible by the shader, while the underlying buffer contents can be 8-, 16- or 32-bit (or even fancier 10- and 11-bit formats). 

I added the code that mapped HLSL buffer types to GLSL layout formats when I was unaware of this difference, and the code always assumes 32-bit formats.  This is a critical issue as actual 32-bit formats are rarely used in practice, 16- or 8-bit being much more common.

Do you have an idea on how to best resolve this? I was thinking we can add support for non-standard attributes, e.g. HLSL code: `[layout(rgba8)] RWBuffer<float4>`. Does that sound like a valid idea?

**Comment by LukasBanana** [comment]
You mean, the actual bit-size is passed to D3D with the respective `DXGI_FORMAT` enumeration, am I right?!
In this case, another input field should do the trick, like the [`VertexSemantic`](https://github.com/LukasBanana/XShaderCompiler/blob/master/inc/Xsc/Xsc.h#L266) structure.

Maybe like this, with two more options:
```cpp
enum class ImageLayoutSize
{
    Bits8,
    Bits16,
    Bits32,
};

struct ImageLayout
{
    std::string name;
    ImageLayoutSize size;
};

struct ShaderOutput
{
    /* ... */
    
    // Either this can be used for individual image layouts ...
    std::vector<ImageLayout> imageLayouts;
    
    // ... or this can be used to set the default bit-size
    ImageLayoutSize defaultImageLayoutSize = ImageLayoutSize::Bits16;
};
```
A corresponding shell command could look like this:
```
-IL, --image-layout [NAME] SIZE
    Specifies the bit-size of the buffer NAME, or default bit-size; valid sizes:
      8, 16, 32
```

I already had an idea of non-standard attributes for a stronger type system, which is currently on standby.
I think this should only be used for experimental features.

**UPDATE:**
I think it's better to do it with the input options, to allow the client programmer to make this layout more dynamic, just like in D3D where the bit-size is not part of the shader, but of the rendering API.

**Comment by BearishSun** [comment]
Yep that's right, DirectX appears to deduce the size from the DXGI_FORMAT, while in GLSL we need to define it in the shader itself.

Doing it as a an output option is problematic for my purposes, but I understand that you don't want non-standard stuff in the compiler. I'll probably go ahead and implement the "attribute" approach locally (I'll probably want a few more non-standard attributes for Banshee so why not get started now).

In particular the problem for me is that shader compilation is very decoupled from the renderer (they happen in separate stages, and execute on separate threads). So at the time of compilation the system has no idea how the shader is meant to be used.

If there is a need to communicate some specifics between the two I use a kind of a meta-language to do so. But I am really unhappy with the meta-language as it often requires me to re-state things already in the actual shader code. In this particular case I'd need to re-state the variable type, identifier and then specify the format. Which is why I prefer the attribute approach.

Thanks for your response.

**Comment by LukasBanana** [comment]
I can well understand this, and as I said, I'm also planning to add a few non-standard attributes.
But they should be opt-in with CMake, like I did with my [`space-attr-extension` branch](https://github.com/LukasBanana/XShaderCompiler/commit/eaf67c8f44c9be41bd08e4d9abf9919f565a27b1).

So just go ahead if this is necessary in your situation, maybe I will add it to the master branch and integrate a pre-processor option via `#ifdef XSC_ENABLE_LANGUAGE_EXT` to allow to disable all language extensions in a project-solution build.

Maybe I implement my proposal with additional options anyway, because both proposals should not interfere.

**Comment by BearishSun** [comment]
Awesome! Should this be merged into `master` or also remain part of a separate branch like `space-attr-extension`?

**Comment by LukasBanana** [comment]
When it's done and works fine I would merge it into master branch.
But I will set those language extensions to be disabled per default in the CMake settings.

**Comment by LukasBanana** [comment]
Should be fixed with [34f7d6e](https://github.com/LukasBanana/XShaderCompiler/commit/34f7d6e4c5b4eb179df62aba8f9141ac79e93538).
